### PR TITLE
SINGA-362 Add functions to support einsum function

### PR DESCRIPTION
**Pull request by sheyujian**
1. add repeat and sum function to support einsum function.
2. all the functions could pass the test case now

### CODE REVIEW

File: src/core/tensor/tensor_math_cpp.h [hunk 1]
```diff
@@ -426,6 +426,20 @@ void Tanh<float, lang::Cpp>(const Tensor& in, Tensor* out,
   TraverseUnary<float>(in, out, tanh_lambda);
 }
 
+template <>
```
Comment (vaanforz) [inline src/core/tensor/tensor_math_cpp.h]: conflict line 429 - 442

File: src/core/tensor/tensor_math.h [hunk 2]
```diff
@@ -251,6 +251,13 @@ void Tanh(const Tensor &in, Tensor *out, Context *ctx) {
   LOG(FATAL) << "Tanh Not Implemented";
 }
 
+/// similar to cudnnTransformTensor
+/// copies the data from one tensor to another tensor with a different layout
+/// the tensors must have the same dimensions but not necessarily the same strides 
+template <typename DType, typename Lang>
```
Comment (vaanforz) [inline src/core/tensor/tensor_math.h]: conflict line 257 - 260

File: include/singa/core/tensor.h [hunk 3]
```diff
@@ -297,6 +309,7 @@ Tensor Sign(const Tensor &in);
 Tensor Sqrt(const Tensor &in);
 Tensor Square(const Tensor &in);
 Tensor Tanh(const Tensor &in);
+Tensor Transform(const Tensor &in);
```
Comment (vaanforz) [inline include/singa/core/tensor.h]: conflict line 312

File: include/singa/core/tensor.h [hunk 4]
```diff
@@ -307,6 +320,7 @@ void Sign(const Tensor &in, Tensor *out);
 void Sqrt(const Tensor &in, Tensor *out);
 void Square(const Tensor &in, Tensor *out);
 void Tanh(const Tensor &in, Tensor *out);
+void Transform(const Tensor &in, Tensor *out);
```
Comment (vaanforz) [inline include/singa/core/tensor.h]: conflict line 323

File: src/core/tensor/tensor.cc [hunk 5]
```diff
@@ -615,6 +753,7 @@ GenUnaryTensorFn(Sign);
 GenUnaryTensorFn(Sqrt);
 GenUnaryTensorFn(Square);
 GenUnaryTensorFn(Tanh);
+GenUnaryTensorFn(Transform);
```
Comment (vaanforz) [inline src/core/tensor/tensor.cc]: conflict line 756

File: src/core/tensor/tensor.cc [hunk 6]
```diff
@@ -122,39 +124,60 @@ void Tensor::ResetLike(const Tensor &in) {
   strides_ = in.strides_;
 }
 
-// if tensor is not transposed yet i.e strides == 1,
-// then we simply change the shape and generate new default strides
-// if tensor is already transposed i.e strides != 1,
-// it should be copied to a new tensor with newly generated default strides
-// TODO(wangwei) raise error if the shape not match
-void Tensor::Reshape(const Shape &shape) {
+Tensor Tensor::Reshape(const Shape &shape) {
```
Comment (vaanforz) [inline src/core/tensor/tensor.cc]: I believe our reshape functions are the same, should be fine to delete either one

File: src/core/tensor/tensor.cc [hunk 7]
```diff
@@ -122,39 +124,60 @@ void Tensor::ResetLike(const Tensor &in) {
...
   shape_ = shape;
   generate_strides();
+  Tensor t(shape, device_, data_type_);
+  t.block_ = block_;
+  t.block_->IncRefCount();
+  return t;
 }
 
-void Tensor::Reshape(Shape &&shape) {
+Tensor Tensor::Reshape(Shape &&shape) {
```
Comment (vaanforz) [inline src/core/tensor/tensor.cc]: I believe our reshape functions are the same, should be fine to delete either one

Comment (sheyujian) [reply src/core/tensor/tensor.cc]: I will delete the conflict lines, is it ok, will it solve the problem?

**Pull request merged**